### DISCUSSION TITLE
Installer: Could not create temporary file in directory "config/params"

### DISCUSSION THREAD

**Discussion by mikeperalta1** [shlinkio/shlink#2292] [Help wanted] [answered]
### Shlink version

shlink4.3.1_php8.2_dist

### PHP version

8.2

### How do you serve Shlink

Self-hosted Apache

### Database engine

MariaDB

### Database version

mysql  Ver 15.1 Distrib 10.5.22-MariaDB, for Linux (x86_64) using  EditLine wrapper

### Current behavior

At the end of running the installer with `vendor/bin/shlink-installer install`, the installer seems to be unable to save the config. I get the following output at the very end:

```
INTEGRATIONS
============

 Do you want to use a redis instance, redis cluster or redis sentinels as a shared cache for Shlink? (recommended if you run a cluster of Shlink instances) (yes/no) [no]:
 > 

 Do you want Shlink to send all visits to an external Matomo server? (yes/no) [no]:
 > 


In RuntimeException.php line 28:
                                                                
  Could not create temporary file in directory "config/params"  
                                                                

install
```

I have no currently made it past this issue so am unable to attempt hosting with apache/php.

### Expected behavior

I expect the config file to be saved, or at least a detailed message explaining the issue. Perhaps even a dump of the file that would be saved, so I can save it.

### Minimum steps to reproduce

RockyLinux 9.5. Download latest distrib zip, unzip in a new user account, run the installer.

I see no indication from audit.log that this request was blocked, am running as the newly created user, and am also able to create/modify files in the unzipped directory area.

**Comment by acelaya** [comment]
This means the system user does not have write permissions in Shlink directory

**Reply by mikeperalta1** [reply, marked as answer]
Thanks, but that wasn't it this time. After researching another post, the issue was my working directory. Apparently, you need to be sitting at install root, not inside the vendor/bin directory. But the program doesn't let you know when it fails. Weird.

But again thanks.

**Reply by acelaya** [reply]
> Apparently, you need to be sitting at install root

Yeah, that's documented here https://shlink.io/documentation/command-line-interface/installation-tool/

![image](https://github.com/user-attachments/assets/a45b85ac-c8ff-4428-a859-b9e6bea01bce)

**Reply by mikeperalta1** [reply]
> > Apparently, you need to be sitting at install root
> 
> Yeah, that's documented here https://shlink.io/documentation/command-line-interface/installation-tool/
> 
> ![image](https://private-user-images.githubusercontent.com/2719332/390336109-a45b85ac-c8ff-4428-a859-b9e6bea01bce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI2OTg2NDIsIm5iZiI6MTczMjY5ODM0MiwicGF0aCI6Ii8yNzE5MzMyLzM5MDMzNjEwOS1hNDViODVhYy1jOGZmLTQ0MjgtYTg1OS1iOWU2YmVhMDFiY2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjdUMDkwNTQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGZkNTdiZTliNTAyMThlMWY3NmVkNzEzYzNlNDEwNDY0NmY1MjFmYmM5MzJhOWEwY2Y1ODFhOGY2ZWIzNWYzNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.KO1Zj-fExUe3c0h2Oi8PcwTiXcfnjzN3l8N7suNgaTE)

It would be nice if the installer gave an indication when that happened. I wasted time hunting for the issue, and I'm sure many others would feel an installer should give meaningful error messages. Would that be possible?

But thank you for the instructions.

**Reply by acelaya** [reply]
Every tool will always have some rough edges. Shlink, and its ecosystem, handle a huge context, and it's impossible to make it perfect at everything and handle absolutely every edge case.

When an issue appears it always seem like an obvious thing to handle in retrospect, but you are not taking into consideration all the things that were in fact considered, handled, tested and automated.

I understand you are frustrated because you wasted your time, but I'm going to strongly encourage you to avoid throwing your frustration towards maintainers which spend their free time working on tools you benefit from for free.

Considering the amount of people using Shlink and how rarely this has been brought up, it has never been a big priority to handle. There's a million other more important things to work on, so I basically just documented this the last time it happened, and this is like the second time it's been reported since forever.

If you think this is so critical to cover, the installer is in https://github.com/shlinkio/shlink-installer. I'm happy to review PRs if you want to give it a go.
